In [23]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import AdaBoostClassifier

<h3>Wczytaj końcowe i przetworzone zbiory danych z zajęć Lab3

In [24]:
df = pd.read_csv('titanic_lab3.csv')

X_test = pd.read_csv('X_test.csv')
X_test.drop('Unnamed: 0', inplace=True, axis=1)

X_train = pd.read_csv('X_train.csv')
X_train.drop('Unnamed: 0', inplace=True, axis=1)

y_test = pd.read_csv('y_test.csv')
y_test.drop('Unnamed: 0', inplace=True, axis=1)

y_train = pd.read_csv('y_train.csv')
y_train.drop('Unnamed: 0', inplace=True, axis=1)

X_train.head(20)

cabin cabinReduced     sex  cabin_map  cabinReduced_map  sex_map
0       NaN            n  female          0                 0        0
1       NaN            n  female          0                 0        0
2       NaN            n  female          0                 0        0
3       NaN            n    male          0                 0        1
4       NaN            n  female          0                 0        0
5       NaN            n    male          0                 0        1
6       E36            E  female         64                 2        0
7       NaN            n  female          0                 0        0
8       C68            C    male        157                 1        1
9       E24            E    male         30                 2        1
10      NaN            n    male          0                 0        1
11      NaN            n  female          0                 0        0
12  C22 C26            C  female          1                 1        0
13      NaN            n    male          0                 0        1
14      NaN            n    male          0                 0        1
15      NaN            n    male          0                 0        1
16      NaN            n    male          0                 0        1
17      NaN            n    male          0                 0        1
18      NaN            n  female          0                 0        0
19      NaN            n    male          0                 0        1

<h3>Scharakteryzuj algorytm Random Forests na podstawie informacji
zawartej w dokumentacji biblioteki scikit-learn dla
RandomForestClassifier. 

<h6>Algorytm ten dopasowuje drzewa decyzyjne do różnych pozbiorów zbioru danych, następnie uśrednia otrzymane wyniki w celu najpeszego dopasowania oraz kotrolowania zjawiska overfittingu. 
    Przyjmuje parametry m.in:
    
    - n_estimators, default=100 - liczba drzew dezyzyjnych
    - criterion{“gini”, “entropy”}, default=”gini” - funkcja mierząca jakość podziału 
    - max_depth, default=None - maksymalna głębokość drzewa
    - min_samples_split, default=2 - minimalna liczba próbek potrzebna do podziału
    - min_samples_leaf, default=1 - minimalna liczba próbek znajdujących się w liściu
    - bootstrap, default=True - czy użyte wartości bootstrapowe czy cały zbiór

<h3>Zbuduj, wytrenuj model oraz dokonaj predykcji używając metody
Random Forest dla danych o wysokiej i niskiej liczbności cech

<h6> wysoka liczebność cech: cabin_map

In [25]:
y_train.head(10)

survived
0         1
1         1
2         1
3         0
4         1
5         0
6         1
7         1
8         0
9         1

In [26]:
y_train.values.ravel()

array([1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1,
       0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0,
       1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0,
       1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1,

In [27]:
Rfc = RandomForestClassifier(n_estimators=220, random_state=42)

In [28]:
Rfc.fit(X_train[['cabin_map', 'sex_map']], y_train.values.ravel())

RandomForestClassifier(n_estimators=220, random_state=42)

In [29]:
predTrain = Rfc.predict_proba(X_train[['cabin_map', 'sex_map']])
predTest = Rfc.predict_proba(X_test[['cabin_map', 'sex_map']].fillna(0))

In [30]:
roc_auc_score(y_test, predTest[:,1])

0.7578120679165975

In [31]:
roc_auc_score(y_train, predTrain[:,1])

0.8535969286350438

<h6> niska liczebność cech: cabinReduced_map

In [32]:
Rfc = RandomForestClassifier(n_estimators=220, random_state=42)
Rfc.fit(X_train[['cabinReduced_map', 'sex_map']], y_train.values.ravel())
predTrain = Rfc.predict_proba(X_train[['cabinReduced_map', 'sex_map']])
predTest = Rfc.predict_proba(X_test[['cabinReduced_map', 'sex_map']].fillna(0))
roc_auc_score(y_test, predTest[:,1])

0.8013245948786019

In [33]:
roc_auc_score(y_train, predTrain[:,1])

0.816719415917359

<h3>Odpowiedz jak różnią się wyniki dla danych z dużą liczebnością i niską
liczebnością cech. Gdzie widoczne jest zbyt duże dopasowanie?
Dlaczego?

<h6> Dla danych z dużą liczebnością cech model jest gorszy od modelu z małą liczebnością cech: 0.75 < 0.80). W modelu pierwszym występuje zbyt duże dopasowanie. Widać to po różnicy wartości oceny modelu pomiędzy zbiorem testowym a treningowym. Dla pierwszego modelu jest to 0.75:0.85. Dla drugiego modelu jest to 0.80:0.82. Im większa różnica pomiędzy tymi wartościami, tym większe zjawisko overfittingu zachodzi. Dużo lepsza sytuacja zachodzi dla modelu drugiego - różnica pomiędzy oceną modelu to zaledwie 0.02.

<h3>Scharakteryzuj algorytm LogisticRegression

<h6>Używany do przewidywania danych binarnych. Jeden z klasyfikatorów liniowych. Wykorzystuje funkcję logiczną (sigmoidalną).
    Przyjmuje parametry m.in.:
    
    - penalty, default="l2" - określa normę kary
    - tol, default=1e-4 - tolerancja na warunek stopu
    - class_weight, default=None - wagi klas
    - solver, default=’lbfgs’ - algorytm użyty do optymalizacji
    - max_iter, default=100 - maksymalna ilość iteracji

<h3>Zbuduj, wytrenuj model oraz dokonaj predykcji używając metody LogisticRegression dla danych o wysokiej i niskiej liczbności cech

<h6> wysoka liczebność cech: cabin_map

In [35]:
Rfc = LogisticRegression(random_state=42)
Rfc.fit(X_train[['cabin_map', 'sex_map']], y_train.values.ravel())
predTrain = Rfc.predict_proba(X_train[['cabin_map', 'sex_map']])
predTest = Rfc.predict_proba(X_test[['cabin_map', 'sex_map']].fillna(0))
roc_auc_score(y_test, predTest[:,1])

0.8011033681765389

In [36]:
roc_auc_score(y_train, predTrain[:,1])

0.8121304826933818

<h6> niska liczebność cech: cabinReduced_map

In [37]:
Rfc = LogisticRegression(random_state=42)
Rfc.fit(X_train[['cabinReduced_map', 'sex_map']], y_train.values.ravel())
predTrain = Rfc.predict_proba(X_train[['cabinReduced_map', 'sex_map']])
predTest = Rfc.predict_proba(X_test[['cabinReduced_map', 'sex_map']].fillna(0))
roc_auc_score(y_test, predTest[:,1])

0.7960151540290914

In [38]:
roc_auc_score(y_train, predTrain[:,1])

0.812497798620301

<h3>Odpowiedz jak różnią się wyniki dla danych z dużą liczebnością i niską liczebnością cech. Gdzie widoczne jest zbyt duże dopasowanie? Dlaczego?

<h6> Dla danych z dużą liczebnością cech oraz dla danych z małą liczebnością cech zachodzą niewielkie różnice pomiędzy modelami. Dla danych testowych model z większą liczebnością cech wypada nieznacznie lepiej niż model z niższą liczebnością cech (0.801 > 0.796). Dla danych treningowych sytuacja natomiast jest odwrotna: model ze zmienną cabin wypada nieznacznie gorzej niż model ze zmienną cabinReduced (0.8121 < 0.8125). Różnice w jednym i drugim modelu są nieznaczne, stąd moim zdaniem można uznać wyniki za wręcz identyczne. Różnice pomiędzy wynikami funcji ROC dla zbiorów testowych oraz treningowych także są bardzo podobne: 0.01 dla dużej kardynalności oraz 0.016 dla małej kardynalności. Z powyższych wyników wynika, że algorytm regresji logicznej jest dość odporny na zjawisko overfittingu.

<h3>Scharakteryzuj algorytm GradientBoosting

<h6>Algorytm polega na korygowaniu parametrów tak, aby osiągnąć jak najlepsze wyniki. Najpierw tworzone jest jedno drzewo decyzyjne. Następnie naprawiamy błędy pierwszego drzewa tworząc kolejne aż do osiągnięcia zadowalających wyników. 
    
    Algorytm GradientBoosting pozwala na optymalizację funkcji strat. Na każdym etapie drzewa regresji dopasowywane są do ujemnego gradientu wielomianowej funkcji strat.
    
    Przyjmuje parametry m.in.:
    
    - loss{‘deviance’, ‘exponential’}, default=’deviance’ - funkcja strat
    - learning_rate, default=0.1 - zmniejsza udział każdego drzewa o learning_rate
    - n_estimators, default=100 - liczba etapów wzmocnienia
    - criterion{‘friedman_mse’, ‘squared_error’, ‘mse’, ‘mae’}, default=’friedman_mse’ -funkcja oceniajaca jakosc podzialu

<h3>Zbuduj, wytrenuj model oraz dokonaj predykcji używając metody GradientBoosting dla danych o wysokiej i niskiej liczbności cech

<h6> wysoka liczebność cech: cabin_map

In [39]:
Rfc = GradientBoostingClassifier(n_estimators=220, random_state=42)
Rfc.fit(X_train[['cabin_map', 'sex_map']], y_train.values.ravel())
predTrain = Rfc.predict_proba(X_train[['cabin_map', 'sex_map']])
predTest = Rfc.predict_proba(X_test[['cabin_map', 'sex_map']].fillna(0))
roc_auc_score(y_test, predTest[:,1])

0.7528621204579393

In [40]:
roc_auc_score(y_train, predTrain[:,1])

0.8631119206597597

<h6> niska liczebność cech: cabinReduced_map

In [41]:
Rfc = GradientBoostingClassifier(n_estimators=220, random_state=42)
Rfc.fit(X_train[['cabinReduced_map', 'sex_map']], y_train.values.ravel())
predTrain = Rfc.predict_proba(X_train[['cabinReduced_map', 'sex_map']])
predTest = Rfc.predict_proba(X_test[['cabinReduced_map', 'sex_map']].fillna(0))
roc_auc_score(y_test, predTest[:,1])

0.8012416348653283

In [42]:
roc_auc_score(y_train, predTrain[:,1])

0.816719415917359

<h3>Odpowiedz jak różnią się wyniki dla danych z dużą liczebnością i niską liczebnością cech. Gdzie widoczne jest zbyt duże dopasowanie? Dlaczego

<h6>Dla zbioru o dużej kardynalności zmiennych model treningowy wypada znacznie lepiej niż model testowy (0.86 > 0.75). Różnica ta (ponad 0.1) ozancza wystepowanie zjawiska overfittingu. Dla zbioru o małej kardynalności zmiennych modele są porównywalnej jakości (0.82 dla treningowego oraz 0.8 dla testowego). Różnica pomiędzy wynikami to zaledwie 0.02 co można interpretować jako brak zjawiska overfittingu. Porównując dużą kardynalność z małą dla zbioru treningowego, znacznie lepiej wypada model z dużą liczbą etykiet (0.86 > 0.81). Dla zbioru testowego sytuacja jednak się odwraca - model z małą ilością etykiet wypada dużo lepiej (0.80 > 0.75). Oznacza to, że model z małą ilością etykiet jest lepszy od modelu z dużą ilością etykiet.

<h3>Scharakteryzuj algorytm AdaBoost

<h6> Algorytm ten zaczyna od dopasowania klasyfikatora do oryginalnego zbioru danych. Nastepnie dopasowuje kopie tego klasyfikatora do tego samego zbioru danych, z tą różnicą, że wagi klas źle dopasowanych zostają zmienione tak, aby kolejne klasyfikatory skupiały się na trudniejszych przypadkach.  
    
Przyjmuje paramtery m.in.:
  
    - base_estimator, default=None - Estymator bazowy
    - n_estimators, default=50 - Maksymalna liczba estymatorów, przy której kończy się wzmacnianie
    - learning_rate, default=1.0 - Waga przypisana do każdego klasyfikatora w każdej iteracji wzmacniającej

<h3>Zbuduj, wytrenuj model oraz dokonaj predykcji używając metody AdaBoost dla danych o wysokiej i niskiej liczbności cech

<h6> wysoka liczebność cech: cabin_map

In [43]:
Rfc = AdaBoostClassifier(n_estimators=220, random_state=42)
Rfc.fit(X_train[['cabin_map', 'sex_map']], y_train.values.ravel())
predTrain = Rfc.predict_proba(X_train[['cabin_map', 'sex_map']])
predTest = Rfc.predict_proba(X_test[['cabin_map', 'sex_map']].fillna(0))
roc_auc_score(y_test, predTest[:,1])

0.785230352303523

In [44]:
roc_auc_score(y_train, predTrain[:,1])

0.8302497245130548

<h6> niska liczebność cech: cabinReduced_map

In [45]:
Rfc = AdaBoostClassifier(n_estimators=220, random_state=42)
Rfc.fit(X_train[['cabinReduced_map', 'sex_map']], y_train.values.ravel())
predTrain = Rfc.predict_proba(X_train[['cabinReduced_map', 'sex_map']])
predTest = Rfc.predict_proba(X_test[['cabinReduced_map', 'sex_map']].fillna(0))
roc_auc_score(y_test, predTest[:,1])

0.7999419279907085

In [46]:
roc_auc_score(y_train, predTrain[:,1])

0.8161759896145195

<h3>Odpowiedz jak różnią się wyniki dla danych z dużą liczebnością i niską liczebnością cech. Gdzie widoczne jest zbyt duże dopasowanie? Dlaczego

<h6>Dla danych z dużą liczebnością cech lepszy jest model trenignowy niż testowy (0.83 > 0.79). Zauważalna różnica (0.04) pomiędzy oceną jakości modeli wskazuje na występowanie zjawiska overfittingu. Dla danych z małą liczebnością cech różnica jest niewielka (0.01). Minimalnie lepiej wypada model treningowy (0.81 > 0.80). Zjawisko overfittingu w tym przypadku nie zachodzi. Lepiej wytrenowany model to model z małą liczbą etykiet (0.80 > 0.79 + brak zjawiska overfittingu).

<h3> Porównaj w/w algorytmy pomiędzy sobą w oparciu o przeprowadzone badania. Które z nich mają największą tendencję do nadmiernego dopasowania?

<h6>Z występujących wyżej algorytmów dość mocno wyróżnia się LogisticRegression – wydaje się on być odporny na zjawisko występowania overfittingu. Zarówno zbiór z dużą kardynalnością zmiennych, jak i małą kardynalnością zmiennych mógłby być uznany za dobry i wystarczający do dalszej predykcji. Nie posiada także paramteru n_estimators. 
    
Pozostałe trzy algorytmy (RandomForestClassifier, GradientBoosting, AdaBoost) dla zbioru z niską kardynalnością osiągają bardzo zbliżone wyniki funkcji roc_auc_score (w okolicach 0.80).  Nie są jednak odporne na zjawisko overfittingu. Największą różnicę w jakości modeli pomiędzy zbiorami treningowym i testowym (a tym samym największe zjawisko overfittingu) widać w algorytmie GradientBoostingClassifier (0.11 różnicy), zaraz za nim znajduje się RandomForest (0.1 różnicy). Na samym końcu z dużo mniejszą różnicą (0.05) jest AdaBoost.  


<h3>Jaka jest różnica pomiędzy predict_proba a predict?

<h6> Predict przewiduje dokładną klasę elementu oraz zwraca jeden element - przewidywaną klasę. Predict_proba przewiduje prawdopodobienstwo nalezenia do jakiejs klasy oraz zwraca listę elementów - prawdopodobieństw przynależenia do danej klasy.